x/config/configurator.go

```diff
@@ -0,0 +1,231 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/internal/decode"
+	errs "go.uber.org/yarpc/internal/errors"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Configurator helps build Dispatchers using runtime configuration.
+//
+// An empty Configurator does not know about any transports. Inform it about
+// the different transports and their configuration parameters using the
+// RegisterTransport function.
+type Configurator struct {
+	knownTransports map[string]*compiledTransportSpec
+}
+
+// New sets up a new empty Configurator. The returned Configurator does not
+// know about any transports. Individual TransportSpecs must be registered
+// against it using the RegisterTransport function.
+func New() *Configurator {
+	return &Configurator{knownTransports: make(map[string]*compiledTransportSpec)}
+}
+
+// RegisterTransport registers a TransportSpec with the given Configurator. An
+// error is returned if the TransportSpec was invalid.
+//
+// If a transport with the same name was already registered, it will be
+// overwritten.
+//
+// Use MustRegisterTransport if you want to panic in case of registration
+// failure.
+func (c *Configurator) RegisterTransport(t TransportSpec) error {
+	if t.Name == "" {
+		return errors.New("name is required")
+	}
+
+	spec, err := compileTransportSpec(&t)
+	if err != nil {
+		return fmt.Errorf("invalid TransportSpec for %q: %v", t.Name, err)
+	}
+
+	// TODO: Panic if a transport with the given name is already registered?
+	c.knownTransports[t.Name] = spec
+	return nil
+}
+
+// MustRegisterTransport registers the given TransportSpec with the
+// Configurator. This function panics if the TransportSpec was invalid.
+func (c *Configurator) MustRegisterTransport(t TransportSpec) {
+	if err := c.RegisterTransport(t); err != nil {
+		panic(err)
+	}
+}
+
+// LoadConfigFromYAML loads a yarpc.Config from YAML. Use LoadConfig if you
+// have your own map[string]interface{} or map[interface{}]interface{} to
+// provide.
+func (c *Configurator) LoadConfigFromYAML(r io.Reader) (yarpc.Config, error) {
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return yarpc.Config{}, err
+	}
+
+	var data map[string]interface{}
+	if err := yaml.Unmarshal(b, &data); err != nil {
+		return yarpc.Config{}, err
+	}
+	return c.LoadConfig(data)
+}
+
+// LoadConfig loads a yarpc.Config from a map[string]interface{} or
+// map[interface{}]interface{}.
+//
+// See the module documentation for the shape the map[string]interface{} is
+// expected to conform to.
+func (c *Configurator) LoadConfig(data interface{}) (yarpc.Config, error) {
```

Comment:
should data has map[string]interface{} type as it is described in comment above?